tensorflow_similarity/architectures/convnext.py

```diff
@@ -0,0 +1,143 @@
+        distance. Defaults to True.
+        include_top: Whether to include the fully-connected layer at the top
+        of the network. Defaults to True.
+        pooling: Optional pooling mode for feature extraction when
+        include_top is False. Defaults to gem.
+        - None means that the output of the model will be the 4D tensor
+          output of the last convolutional layer.
+        - avg means that global average pooling will be applied to the
+          output of the last convolutional layer, and thus the output of the
+          model will be a 2D tensor.
+        - max means that global max pooling will be applied.
+        - gem means that global GeneralizedMeanPooling2D will be applied.
+          The gem_p param sets the contrast amount on the pooling.
+        gem_p: Sets the power in the GeneralizedMeanPooling2D layer. A value
+        of 1.0 is equivalent to GlobalMeanPooling2D, while larger values
+        will increase the contrast between activations within each feature
+        map, and a value of math.inf will be equivalent to MaxPool2d.
+    """
+    inputs = layers.Input(shape=input_shape)
+    x = inputs
+
+    if variant not in CONVNEXT_ARCHITECTURE:
+        raise ValueError("Unknown ConvNeXt variant. Valid TINY BASE LARGE SMALL XLARGE")
+
+    x = build_convnext(variant, weights, trainable)(x)
+
+    if pooling == "gem":
+        x = GeneralizedMeanPooling2D(p=gem_p, name="gem_pool")(x)
+    elif pooling == "avg":
+        x = layers.GlobalAveragePooling2D(name="avg_pool")(x)
+    elif pooling == "max":
+        x = layers.GlobalMaxPooling2D(name="max_pool")(x)
+
+    if include_top and pooling is not None:
+        if l2_norm:
+            outputs = MetricEmbedding(embedding_size)(x)
+        else:
+            outputs = layers.Dense(embedding_size)(x)
+    else:
+        outputs = x
+
+    return SimilarityModel(inputs, outputs)
+
+
+def build_convnext(variant: str, weights: str | None = None, trainable: str = "full") -> tf.keras.Model:
+    """Build the requested ConvNeXt
+
+    Args:
+        variant: Which Variant of the ConvNeXt to use.
+        weights: Use pre-trained weights - the only available currently being
+        imagenet.
+        trainable: Make the ConvNeXt backbone fully trainable or partially
+        trainable.
+        - "full" to make the entire backbone trainable,
+        - "partial" to only make the last 3 block trainable
+        - "frozen" to make it not trainable.
+    Returns:
+        The output layer of the convnext model
+    """
+    convnext_fn = CONVNEXT_ARCHITECTURE[variant.upper()]
+    convnext = convnext_fn(weights=weights, include_top=False)
+
+    if trainable == "full":
+        convnext.trainable = True
+    elif trainable == "partial":
+        convnext.trainable = True
+        for layer in convnext.layers:
+            # freeze all layeres befor the last 3 blocks
+            if not re.search("^block[5,6,7]|^top", layer.name):
```

Comment:
I'm also trying out this architecture. But does this EfficientNetV2 layer naming apply to convnext?

```python
model = tf.keras.applications.ConvNeXtBase()
[l.name for l in model.layers if re.search("^block[5,6,7]|^top", l.name)]
# this outputs []
```

The [test](https://github.com/tensorflow/similarity/pull/356/files#diff-2a82cf98e235f66b2d31d8e65f7fd517855fd71348af8de04f6da6371ac7f34aR48) also suggests `partial` is not being applied as expected since the number of trainable layers is 0 with partial.